### DISCUSSION TITLE
Unable to set Custom User Profile fields in login script

### DISCUSSION THREAD

**Discussion by ryanr-gpn** [OrchardCMS/OrchardCore#16740] [Q&A]
I am using Microsoft Entra ID for user login/registration. I am trying to populate some custom user profile fields (e.g. first and last name) from the external claims that come back from the external login. Ideally I'd like to do this during registration but have not found a way to do that. So I'm using the User Login script and setting the "propertiesToUpdate" field, but it's not actually updating the user.  I have a UserProfile content type (with CustomUserSettings for the stereotype) and I can see those fields in the admin, but they do not get updated when logging in. I've tried 
`context.propertiesToUpdate={
      	"UserProfile":{
        	"UserProfile": {
				"FirstName": { "Text": "MyFirst" },
				"LastName": { "Text": "MyLast" }
			}
        }
    };`
and a few variations on that. Am I going about this the right way, or is there a better way to accomplish it? I'm not seeing errors - it's just not updating the user record.

**Comment by Piedone** [comment]
Since I've never used that I don't know a solution from the top of my head, but how about trying it with an `IExternalLoginEventHandler` from C# instead?

**Comment by gvkries** [comment]
IIRC the script should look something like this:

```javascript
context.propertiesToUpdate = {
    UserProfile: {
        ContentType: "UserProfile",
        UserProfile: {
            FirstName: {
                Text: "MyFirstname"
            }
        }
    }
};
```

**Comment by ryanr-gpn** [comment]
@gvkries that didn't seem to work either. Maybe it's because the UserProfile property is not getting initially created? Looking at the record in the database I just see "Properties":{} for that newly created user.

If I look at a user record that I manually updated in the admin, I see "Properties":{"UserProfile":{"ContentItemId":"4mj5q1j975cnew61evyar7mpdd","ContentItemVersionId":null,"ContentType":"UserProfile","DisplayText":"","Latest":false,"Published":false,"ModifiedUtc":"2024-09-19T13:22:10.5572826Z","PublishedUtc":null,"CreatedUtc":null,"Owner":"","Author":"user+company.com","UserProfile":{"FirstName":{"Text":"first"},"LastName":{"Text":"last"}}}}

**Reply by gvkries** [reply]
Works for me. Are you using v2?

**Reply by ryanr-gpn** [reply]
No, I'm on 1.8.4. Does this only work on v2?
Ideally I'd like to do this during the initial user registration but I'm not seeing a way to do that. Do you know if that's possible?
Thanks for your help.

**Reply by gvkries** [reply]
Yes, this feature is new in 2.0. You would have to use custom code in 1.8 like @Piedone suggested.

**Comment by kevinaig** [comment]
Were you able to get a working solution to bring in Entra ID properties into the user profile?

**Reply by twhug** [reply]
This is a bit of a necro but since the question is still unanswered - using @gvkries' answer and using v2.1.4, we're correctly able to set custom user properties pulled from EntraID as part of the user login script. In our case, we're populating a simple 'first name' field from the following claim:
_http://schemas.xmlsoap.org/ws/2005/05/identity/claims/givenname_

However, from my testing, it looks like this only works if the custom property is already initialised, e.g. by setting a value through the web UI. 

If the property has never had a value saved for a given user, then the 'propertiesToUpdate' call throws an exception when that user logs in.

It would be amazing if there's a workaround, or if a function can be added that would allow us to push an already-existing field to a user's Properties.

**Reply by gvkries** [reply]
I'm not aware of an bug in this functionality, can you share the script and the exception?

**Reply by twhug** [reply]
> I'm not aware of an bug in this functionality, can you share the script and the exception?

Sorry, I never properly responded to this comment. 

Right now, I appear to be having issues actually passing anything across - not sure how I ever had it working.

```
switch (context.loginProvider) {
    case "AzureAd":
        context.externalClaims.forEach(claim => {
            if  (claim.type === "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/givenname") {
                userFirstName = claim.value;
            }
        });

        context.propertiesToUpdate = {
                gUserProfile: {
                    ContentType: "gUserProfile",
                    gUserProfile: {
                        UserShortName: {
                            Text: userFirstName
                        }
                    }
                }
            };  
}
```

This throws the following exception:

> Unexpected error while updating a user. System.InvalidOperationException: An object with the same identity is already part of this transaction. Reload it before doing any changes on it.

Nothing else in the script should be modifying the user e.g. I've removed any other 'claimsToUpdate' and 'propertiesToUpdate' calls that were there originally. Still on 2.1.4.

**Reply by gvkries** [reply]
Your script looks good to me and the exception indicates that an unexpected user instance is updated. Is this error happening when a new user registers only, or on every login?

**Reply by twhug** [reply]
It's every login, sadly.

**Reply by twhug** [reply]
@gvkries I believe this may be an issue with YesSql under 2.1.4 - did some preliminary testing using 2.2.1, the exception is no longer thrown and it appears as though the setting is updated as expected. Will do some more testing to confirm but looking good right now...

**Reply by gvkries** [reply]
That's very good, thank you.
